rcl_yaml_param_parser/test/test_parser.cpp

```diff
@@ -0,0 +1,351 @@
+  EXPECT_EQ(rcl_yaml_node_struct_init(allocator), nullptr);
+
+  // Bad alloc of params_st->node_names
+  set_time_bomb_allocator_calloc_count(allocator, 1);
+  EXPECT_EQ(rcl_yaml_node_struct_init(allocator), nullptr);
+
+  // Bad alloc of params_st->params
+  set_time_bomb_allocator_calloc_count(allocator, 2);
+  EXPECT_EQ(rcl_yaml_node_struct_init(allocator), nullptr);
+
+  // Check this doesn't die.
+  rcl_yaml_node_struct_fini(nullptr);
+}
+
+TEST(RclYamlParamParser, node_copy) {
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rcl_params_t * params_st = rcl_yaml_node_struct_init(allocator);
+  EXPECT_NE(params_st, nullptr);
+
+  EXPECT_EQ(nullptr, rcl_yaml_node_struct_copy(nullptr));
+
+  const char node_name[] = "node name";
+  const char param_name[] = "param name";
+  const char yaml_value[] = "true";
+  EXPECT_TRUE(rcl_parse_yaml_value(&node_name[0], &param_name[0], &yaml_value[0], params_st));
+
+  rcl_params_t * copy = rcl_yaml_node_struct_copy(params_st);
+  EXPECT_NE(copy, nullptr);
+  rcl_yaml_node_struct_fini(copy);
+
+  params_st->allocator = get_time_bomb_allocator();
+
+  // init of out_params_st fails
+  set_time_bomb_allocator_calloc_count(params_st->allocator, 0);
+  EXPECT_EQ(nullptr, rcl_yaml_node_struct_copy(params_st));
+
+  set_time_bomb_allocator_calloc_count(params_st->allocator, 1);
+  EXPECT_EQ(nullptr, rcl_yaml_node_struct_copy(params_st));
+
+  constexpr int expected_num_calloc_calls = 5;
+  for (int i = 0; i < expected_num_calloc_calls; ++i) {
+    // Check various locations for allocation failures
+    set_time_bomb_allocator_calloc_count(params_st->allocator, i);
+    EXPECT_EQ(nullptr, rcl_yaml_node_struct_copy(params_st));
+  }
+  // Check that the expected number of calloc calls occur
+  set_time_bomb_allocator_calloc_count(params_st->allocator, expected_num_calloc_calls);
+  copy = rcl_yaml_node_struct_copy(params_st);
+  EXPECT_NE(nullptr, copy);
+  rcl_yaml_node_struct_fini(copy);
+
+  // Reset calloc countdown
+  set_time_bomb_allocator_calloc_count(params_st->allocator, -1);
+
+  constexpr int expected_num_malloc_calls = 3;
+  for (int i = 0; i < expected_num_malloc_calls; ++i) {
+    set_time_bomb_allocator_malloc_count(params_st->allocator, i);
+    EXPECT_EQ(nullptr, rcl_yaml_node_struct_copy(params_st));
+  }
+
+  // Check that the expected number of malloc calls occur
+  set_time_bomb_allocator_malloc_count(params_st->allocator, expected_num_malloc_calls);
+  copy = rcl_yaml_node_struct_copy(params_st);
+  EXPECT_NE(nullptr, copy);
+  rcl_yaml_node_struct_fini(copy);
+
+  constexpr int num_malloc_calls_until_copy_param = 2;
+
+  // Check integer value
+  int64_t temp_int = 42;
+  set_time_bomb_allocator_malloc_count(params_st->allocator, num_malloc_calls_until_copy_param);
+  params_st->params->parameter_values[0].bool_value = nullptr;
+  params_st->params->parameter_values[0].integer_value = &temp_int;
+  EXPECT_EQ(nullptr, rcl_yaml_node_struct_copy(params_st));
```

Comment:
if the time bomb allocator is set to `num_malloc_calls_until_copy_param`, why is this returning `nullptr`? I would expect to get the parameter copy

Reply:
Because that's how many malloc calls occur before it attempts to copy, which requires a malloc. So it should fail after that many calls.